Installation script trac 11199

manics opened the pull request:
Download/clone the omero_searcher repo somewhere. Enter the directory, run `./install.sh` to see a help message.

Run `./install.sh /path/to/omero/server` to install OMERO.searcher. This will also attempt to install the Python dependencies using pip, however there is also a dependency on the `freeimage` library which can't be installed automatically. For your sanity you might want to create a virtualenv first.

This script also attempts to configure the `omero.web.apps` config value, however it does not create the OMERO.searcher/PySLID data directory- you must do this manually.

manics committed:
Added installation script

manics committed:
Add note about creating data directory

manics committed:
Attempt to install Python dependencies

manics commented:
Only remaing question is does it matter that the PySLID Python module is installed in the system site-packages directory, even though it has a hard dependency on the omero python module which isn't by default in the Python search path? Or shall we just ignore this for now.

joshmoore commented on install.sh:
```diff
@@ -0,0 +1,118 @@
+
+echo "Installing scripts"
+mkdir "$SCRIPT_DEST"
+cp -a scripts/* "$SCRIPT_DEST"
```
This assumes it will always be run via `sh install.sh` without a relative path, right?

manics replied:
Yes... if you think it could cause problems I can change it.

mtbc replied:
Can't use `$0` in case it was run with `source`, nor `$BASH_SOURCE` in case they're not using bash. ):

joshmoore commented:
@manics: the installation of OMERO.py is a good point, but probably outside the scope of this. I don't know if you'd like to modify PSLID to produce a warning if it's imported without OMERO being on `PYTHONPATH`, but that would be one possibility. In the mid-term, we should really get OMERO.py pip-installable.

mtbc commented:
I wasn't quite sure what to make of `/path/to/omero/server/`. The trailing separator doesn't actually seem to be required. I'm assuming that it's the same as what I put in `OMERO_PREFIX` rather than the path of the actual `omero` python script executable or something.

mtbc commented:
Oh, I die with,

```
creating /Library/Python/2.7/site-packages/mahotas

error: could not create '/Library/Python/2.7/site-packages/mahotas': Permission denied
```

Was I meant to run it as root? (Hopefully not.) Maybe that's what the virtualenv stuff is about, I never did make time to read about that. Will keep investigating...

manics commented:
`/path/to/omero/server/`
I was trying to make it clear that it's expecting a directory and not some random file somewhere... suggestions for making it clearer? I've never set `OMERO_HOME` or `OMERO_PREFIX` for anything.

Python error: You need to install the python modules into the system path, so either as root, or install Python with homebrew (which basically changes the permissions so everything is owned by you) or use virtualenv which creates a mini Python sandbox/environment owned by you, and has the added benefit of not destroying your python installation when you accidentally overwrite a module with an incompatible version.

manics commented:
>  I don't know if you'd like to modify PSLID to produce a warning if it's imported without OMERO being on PYTHONPATH

`pyslid/__init__.py` currently imports everything including the OMERO dependent submodules, so an import will currently fail anyway.

mtbc commented:
Okay, I have now got virtualenv going and I seem to have a new Searcher tab in the web client that kind of works. (I don't see how to cause feature calculation, but that's by the by.) So, I guess that this PR is **good to merge**.

Yes, the trailing slash is a hint. `OMERO_PREFIX` is suggested a few times in our sysadmin and developer documentation so it might be a useful thing to refer to; I simply needed `./install.sh $OMERO_PREFIX`. It needn't block this PR, but an example of a likely value for it might go a long way in showing what it isn't.

mtbc commented:
Maybe it would be good to ensure that some of the above discussion is ticketed: PySLID warning well if OMERO.py isn't importable, making the latter pip-installable, whatever.

manics commented:
```
>>> import pyslid
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/pyslid/__init__.py", line 1, in <module>
    import features
  File "/usr/local/lib/python2.7/site-packages/pyslid/features.py", line 86, in <module>
    import omero, pyslic, pyslid.utilities, pyslid.image
ImportError: No module named omero
>>>
```

Ugly, but it's pretty clear!

Ticket for OMERO.py: https://trac.openmicroscopy.org.uk/ome/ticket/11216

mtbc commented:
Excellent, thank you.

manics commented:
> I don't see how to cause feature calculation, but that's by the by.

Actually it's kind of important, since it's part of the installation :). There should be a `searcher` scripts submenu.

mtbc commented:
Oh, so there is! Thank you. Given that they are already on the Searcher menu, I am not sure each submenu item needs an "Omero Searcher" prefix.

If I try feature calculation, the script runs happily (though its report mentions only the first of the images I selected, so maybe only one of them got any features), but then search attempts always report, "The ContentDB for feature-set slf33 could not be found. Have you calculated any features?" Whether that's an installation issue or something else, I've no idea, but I don't seem to be able to calculate features such that I thereby change subsequent behavior.

mtbc commented:
Huh, I've also achieved a,

```
Traceback (most recent call last):
  File "./script", line 386, in <module>
    runScript()
  File "./script", line 374, in runScript
    message += processImages(client, scriptParams) + '\n'
  File "./script", line 224, in processImages
    channels, zselect, tselect, recalc, disableCdb)
  File "./script", line 171, in extractFeatures
    disableCdb)
  File "./script", line 68, in extractFeaturesOneChannel
    pixels=0, channel=channels[0], zslice=zslice, timepoint=timepoint)
  File "/tmp/ve/lib/python2.7/site-packages/pyslid/features.py", line 724, in link
    table.initialize(columns)
AttributeError: 'NoneType' object has no attribute 'initialize'
```

Nothing in `pyslid.data/`

manics commented:
> I am not sure each submenu item needs an "Omero Searcher" prefix.

Added to description in [Trac 10757](https://trac.openmicroscopy.org.uk/ome/ticket/10757).

Traceback error: Note added to [Trac 10464](https://trac.openmicroscopy.org.uk/ome/ticket/10464), seems to be an ongoing issue with Ice, Tables and Scripts.

I'm going to use this installation script on a CI server, which should help iron out any other problems.

manics committed:
Remove most dependencies from requirements.

Dependencies should be specified in PySLID setup.py instead, to prevent duplication of requirements in omero_searcher and PySLID.

manics committed:
Use git://github.com/manics/pyslid.git@merge_all instead
This branch is created using scc merge --default all master

manics committed:
Use OMERO_SERVER instead of an example path

manics committed:
Add XXXXXX to mktemp argument

manics commented:
I've added a few additional commits to fix problems- this script is now being run as part of the manually triggered OMERO.searcher CI builds: http://hudson.openmicroscopy.org.uk/view/Analysis/job/ANALYSIS-OMERO-SEARCHER-merge/ 

If feature calculation isn't working properly for you then when you're back I could sit down with you to figure out whether its likely to be a problem for others.

mtbc commented:
My "If I try feature calculation ..." paragraph above still seems to hold. Would you like to come and look?

mtbc commented:
My problem turned out to be,

```
$ python
Python 2.7.2 (default, Oct 11 2012, 20:14:37) 
[GCC 4.2.1 Compatible Apple Clang 4.0 (tags/Apple/clang-418.0.60)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import tables
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/src/tables/tables/__init__.py", line 30, in <module>
    from tables.utilsExtension import getPyTablesVersion, getHDF5Version
ImportError: dlopen(/usr/local/src/tables/tables/utilsExtension.so, 2): Library not loaded: /usr/local/lib/libhdf5.7.dylib
  Referenced from: /usr/local/src/tables/tables/utilsExtension.so
  Reason: image not found
>>> 
```

which isn't caused by this PR.

manics commented:
As discovered by Mark OMERO.searcher assumes OMERO.tables is already installed and working.

pwalczysko commented:
Discovered just that you have to now do `sudo pip install virtualenv` and activate it, using `[petr@ls25165 ~]$  . /Users/petr/searcherenv/bin/activate`.
Then do pip install numpy 
pip install scipy
and then hope for the best.
Not getting to the end of this.

pwalczysko commented:
Managed to run the script to the end. No apparent errors. Created a directory under my /OMERO/folder named pyslid.data. 
Started my local server and my local web server. This went fine.
Then went to the web browser, and got 

```

Environment:


Request Method: GET
Request URL: http://localhost:4080/

Django Version: 1.3.1
Python Version: 2.7.2
Installed Applications:
['django.contrib.staticfiles',
 'django.contrib.markup',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.sites',
 'omeroweb.feedback',
 'omeroweb.webadmin',
 'omeroweb.webclient',
 'omeroweb.webgateway',
 'omeroweb.webtest',
 'omeroweb.webredirect',
 'omeroweb.webstart',
 u'omeroweb.omero_searcher']
Installed Middleware:
('django.middleware.common.CommonMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware')


Traceback:
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/core/handlers/base.py" in get_response
  111.                         response = callback(request, *callback_args, **callback_kwargs)
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/views/generic/base.py" in view
  47.             return self.dispatch(request, *args, **kwargs)
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/views/generic/base.py" in dispatch
  68.         return handler(request, *args, **kwargs)
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/views/generic/base.py" in get
  151.         url = self.get_redirect_url(**kwargs)
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/views/generic/base.py" in get_redirect_url
  140.         if self.url:
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/utils/functional.py" in __wrapper__
  193.                 res = self.__func(*self.__args, **self.__kw)
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/core/urlresolvers.py" in reverse
  391.             *args, **kwargs)))
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/core/urlresolvers.py" in reverse
  312.         possibilities = self.reverse_dict.getlist(lookup_view)
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/core/urlresolvers.py" in _get_reverse_dict
  229.             self._populate()
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/core/urlresolvers.py" in _populate
  208.                     for name in pattern.reverse_dict:
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/core/urlresolvers.py" in _get_reverse_dict
  229.             self._populate()
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/core/urlresolvers.py" in _populate
  197.         for pattern in reversed(self.url_patterns):
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/core/urlresolvers.py" in _get_url_patterns
  279.         patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/core/urlresolvers.py" in _get_urlconf_module
  274.             self._urlconf_module = import_module(self.urlconf_name)
File "/Users/petr/Work/openmicroscopy/dist/lib/python/django/utils/importlib.py" in import_module
  35.     __import__(name)
File "/Users/petr/Work/openmicroscopy/dist/lib/python/omeroweb/omero_searcher/urls.py" in <module>
  16. from omeroweb.omero_searcher import views
File "/Users/petr/Work/openmicroscopy/dist/lib/python/omeroweb/omero_searcher/views.py" in <module>
  27. import pyslid

Exception Type: ImportError at /
Exception Value: No module named pyslid
```

![screen shot 2013-07-24 at 18 48 46](https://f.cloud.github.com/assets/2478303/850476/5328a89c-f489-11e2-9011-595d121f9906.png)

manics commented:
@pwalczysko did you run omero admin start and omero web start from within the virtualenv?

pwalczysko commented:
No

manics commented:
You'll need to, otherwise omero won't see the new python modules.

pwalczysko commented:
Done and Web starts with Searcher tab in it. 
But after the Feature calculation script was run on a dataset, got an error.

```
No handlers could be found for logger "omero.gateway"
Traceback (most recent call last):
  File "./script", line 386, in <module>
    runScript()
  File "./script", line 374, in runScript
    message += processImages(client, scriptParams) + '\n'
  File "./script", line 240, in processImages
    channels, zselect, tselect, recalc, disableCdb)
  File "./script", line 171, in extractFeatures
    disableCdb)
  File "./script", line 52, in extractFeaturesOneChannel
    pixels, channels, zslice, timepoint, debug=True)
  File "/Users/petr/searcherenv/lib/python2.7/site-packages/pyslid/features.py", line 249, in calculate
    img.channeldata[ 'protein' ] = scipy.misc.imresize( plane, scale )
AttributeError: 'module' object has no attribute 'imresize'
```

pwalczysko commented:
In order to make the python environment to see the python modules necessary for OMERO, create your environment with the option `--system-site-packages`. This means for example

```

[petr@ls25165 ~/Work/omero_searcher]$ virtualenv --system-site-packages searcherenv
```

After this, activate your environment (see above) and install PIL, for example:

```
(searcherenv)[petr@ls25165 ~/Work/omero_searcher]$ pip install PIL
```

After this, still being in the environment ("searcherenv" in these examples), restart your omero server and start your web server.

pwalczysko commented:
After doing all the above, I have managed to start a web server, with running searcher in it. The feature calculation script was successfully run on a dataset. A search was performed. The results were okay. 

I think that the PR basically works and does what it should do. 
What would be crucial now, is to go over the comments in this PR and collate the information gathered during the testing to have a more streamlined workflow - possibly some Readme.txt which would lead the tester/user through all the Python environment trappings and silly mistakes.

manics committed:
Check for PIl, numpy, scipy. Additional install instructions

manics commented:
@pwalczysko Added additional checks for Python modules, add note that OMERO.tables is required.

pwalczysko commented:
Everything still works as expected. 
The changes in the install.sh are there - would it make sense to put there also the python virtualenv thing ? 
Otherwise ready to merge.

manics commented:
I'm making the assumption that people using virtualenv are familiar with it, and that once this goes into production people will just use whatever setup they've got for their existing server. i.e. If they already run OMERO.server in a virtualenv then there's no problem, if they run it using system packages they can continue to do so. The `--system-site-packages` flag makes things easier to setup, but I'm not sure whether we should recommend it as good practice.

@joshmoore Do we need some development/environment guidelines for the OME consortium addons?

joshmoore commented:
It certainly couldn't hurt, @manics. We probably have something close to this under the contrib section which @hflynn is currently looking at, and @sbesson has certainly played with various ways of getting Python and modules setup, some of which we certainly _wouldn't_ suggest (at least on OSX) which we should get written down. But in general, I agree with you: whether or virtualenv or system python is up to the "user", and they'll just need to understand which situation they're under. If you and @sbesson think we should suggest `system-site-packages` across the board, we can definitely do that.

hflynn commented:
@manics the contributing section is currently https://www.openmicroscopy.org/site/support/omero4/developers/contributing.html - https://www.openmicroscopy.org/site/support/omero4/developers/policies.html may be the most obvious place to add something if you think it's appropriate. Either that or add a new file to this section.

manics commented:
I'll leave the virtualenv flags out of the install script instructions until we decide.

Ticket added https://trac.openmicroscopy.org.uk/ome/ticket/11328

manics commented:
Is everyone happy with this?

joshmoore commented:
No objections from me.

manics commented:
Happy to merge then? #13 needs to modify the install script (adds a templatetags directory).
